File: src/main/java/ch/softappeal/yass/core/remote/session/Packet.java
package ch.softappeal.yass.core.remote.session;

import ch.softappeal.yass.core.remote.Message;
import ch.softappeal.yass.util.Check;
import ch.softappeal.yass.util.Nullable;

import java.io.Serializable;

/**
 * A {@link Message} with a request number.
 */
public final class Packet implements Serializable {

  private static final long serialVersionUID = 1L;

  private final int requestNumber;
  @Nullable private final Message message;

  public boolean isEnd() {
    return requestNumber == END_REQUEST_NUMBER;
  }

  private void checkNotEnd() {
    if (isEnd()) {
      throw new IllegalStateException("not allowed if isEnd");
    }
  }

  /**
   * Must not be called if {@link #isEnd()}.
   */
  public int requestNumber() {
    checkNotEnd();
    return requestNumber;
  }

  /**
   * Must not be called if {@link #isEnd()}.
   */
  public Message message() {
    checkNotEnd();
    return message;
  }

  /**
   * @param requestNumber must not be {@link #END_REQUEST_NUMBER}; use {@link #END} instead
   */
  public Packet(final int requestNumber, final Message message) {
    if (requestNumber == END_REQUEST_NUMBER) {
      throw new IllegalArgumentException("use END");
    }
    this.requestNumber = requestNumber;
    this.message = Check.notNull(message);
  }

  private Packet() {
    requestNumber = END_REQUEST_NUMBER;
    message = null;
  }

  public static boolean isEnd(final int requestNumber) {
    return requestNumber == END_REQUEST_NUMBER;
  }

  public static final Packet END = new Packet();

  public static final int END_REQUEST_NUMBER = 0;

}

File: kotlin/yass/main/ch/softappeal/yass/remote/session/Packet.kt
package ch.softappeal.yass.remote.session

import ch.softappeal.yass.remote.Message

const val EndRequestNumber = 0

fun isEndPacket(requestNumber: Int): Boolean =
    requestNumber == EndRequestNumber

val EndPacket = Packet()

class Packet {
    private val _requestNumber: Int
    private val _message: Message?

    val isEnd: Boolean
        get() = isEndPacket(_requestNumber)

    val requestNumber: Int
        get() {
            check(!isEnd)
            return _requestNumber
        }

    val message: Message
        get() = _message!!

    constructor(requestNumber: Int, message: Message) {
        require(!isEndPacket(requestNumber))
        _requestNumber = requestNumber
        _message = message
    }

    internal constructor() {
        _requestNumber = EndRequestNumber
        _message = null
    }
}
